practice/tic_tac_toe/list_board.py

```diff
@@ -0,0 +1,101 @@
+        self._rows = [
+            [' ', ' ', ' '],
+            [' ', ' ', ' '],
+            [' ', ' ', ' ']
+        ]
+
+    def __repr__(self):
+        """Implement repr() function.
+
+        >>> ListTTTBoard()
+        ListTTTBoard [[' ', ' ', ' '], [' ', ' ', ' '], [' ', ' ', ' ']]
+        """
+        return 'ListTTTBoard {}'.format(self._rows)
+
+    def __eq__(self, other):
+        """Implement the equality function.
+
+        >>> ListTTTBoard() == ListTTTBoard()
+        True
+        >>> a = ListTTTBoard()
+        >>> b = ListTTTBoard()
+        >>> a._rows[0][0] = 'X'
+        >>> a == b
+        False
+        """
+        return self._rows == other._rows
+
+    def place_token(self, x, y, token):
+        """Place a token at x, y location (0, 0 is upper left).
+
+        >>> a = ListTTTBoard()
+        >>> a.place_token(1, 1, 'X')
+        >>> a._rows
+        [[' ', ' ', ' '], [' ', 'X', ' '], [' ', ' ', ' ']]
+        """
+        self._rows[y][x] = token
+
+    def calc_winner(self):
+        """Return the token string that has a row, column or diagonal."""
+        _winner = self._calc_winner_row_col()
+        if _winner is not None:
+            return _winner
+        _winner = self._calc_winner_diag()
+        if _winner is not None:
+            return _winner
+
+    def _calc_winner_row_col(self):
+        """Return the token string if it occurs in an entire column.
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['X', 'O', 'O'], ['X', 'O', ' ']]
+        >>> board._calc_winner_row_col()
+        'X'
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['O', 'O', 'O'], ['X', 'O', ' ']]
+        >>> board._calc_winner_row_col()
+        'O'
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['O', ' ', 'O'], ['X', 'O', ' ']]
+        >>> board._calc_winner_row_col()
+        """
+        for token in TOKENS:
+            for i in range(3):
+                if (all([self._rows[i][j] == token for j in range(3)]) or
+                   all([self._rows[j][i] == token for j in range(3)])):
+                    return token
+
+    def _calc_winner_diag(self):
+        """Return the token string if it occurs in a positive diagonal.
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['O', 'X', 'O'], ['X', 'O', 'X']]
+        >>> board._calc_winner_diag()
+        'X'
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['O', 'O', 'O'], ['O', 'O', 'X']]
+        >>> board._calc_winner_diag()
+        'O'
+        >>> board = ListTTTBoard()
+        >>> board._rows = [['X', ' ', 'O'], ['O', ' ', 'O'], ['X', 'O', ' ']]
+        >>> board._calc_winner_diag()
+        """
+        for token in TOKENS:
+            if (all([self._rows[i][i] == token for i in range(3)]) or
+               all([self._rows[2 - i][i] == token for i in range(3)])):
+                return token
```

Comment:
I am curious:  You replaced the structure _rows[2-i][i] with _rows[-i][i] as it iterates through range(3).  Does this work the same?  It appears to me that when i = 0, my version gives _row[2][0] but the new one gives _row[-0][0].  This does not fit my model of how it should work unless I am missing something key.